### PR TITLE
refactor: remove lodash, add fast-deep-equal

### PR DESCRIPTION
**Pull request by JamieMagee**
With the help of [You don't need lodash](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore). Exceptions are:

- Replaced _.isEqual with [fast-deep-equal](https://github.com/epoberezkin/fast-deep-equal)
- Reimplemented sampleSize
- Reimplemented camelCase

Done as part of my investigation into jest memory leakage. Unfortunately it's brought in via a lot of our dependencies, but it helps a bit.

### CODE REVIEW

File: lib/versioning/ruby/version.ts [hunk 1]
```diff
@@ -25,7 +24,7 @@ const floor = (version: string): string =>
 // istanbul ignore next
 const incrementLastSegment = (version: string): string => {
   const segments = create(version).release().getSegments();
-  const nextLast = parseInt(last(segments), 10) + 1;
+  const nextLast = parseInt(segments[segments.length - 1], 10) + 1;
```
Comment (viceice) [inline lib/versioning/ruby/version.ts:27]:
```suggestion
  const nextLast = parseInt(segments.pop(), 10) + 1;
```

File: lib/versioning/ruby/version.ts [hunk 2]
```diff
@@ -25,7 +24,7 @@ const floor = (version: string): string =>
...
 
   return [...segments.slice(0, -1), nextLast].join('.');
```
Comment (viceice) [inline lib/versioning/ruby/version.ts:29]:
```suggestion
  return [...segments, nextLast].join('.');
```

**Review by viceice — CHANGES_REQUESTED** [review]
needs lint fixes too

**Comment by JamieMagee** [comment]
> needs lint fixes too

Yeah, I realised just after I sent it out that I'd been testing with just `yarn jest`

File: lib/workers/pr/index.ts [hunk 3]
```diff
@@ -524,3 +522,19 @@ export async function checkAutoMerge(
   logger.debug('No automerge');
   return false;
 }
+
+function sampleSize(array: string[], n: number) {
```
Comment (rarkins) [inline lib/workers/pr/index.ts:526]: maybe these types of function can be moved into somewhere in `lib/util` ? `sampleSize`, `camelCase`, etc?

File: yarn.lock [hunk 4]
```diff
@@ -6376,7 +6381,7 @@ lodash.without@~4.4.0:
   resolved "https://registry.yarnpkg.com/lodash.without/-/lodash.without-4.4.0.tgz#3cd4574a00b67bae373a94b748772640507b7aac"
   integrity sha1-PNRXSgC2e643OpS3SHcmQFB7eqw=
 
-lodash@4.17.15, lodash@^4.17.11, lodash@^4.17.13, lodash@^4.17.14, lodash@^4.17.15, lodash@^4.17.4:
+lodash@^4.17.11, lodash@^4.17.13, lodash@^4.17.14, lodash@^4.17.15, lodash@^4.17.4:
```
Comment (rarkins) [inline yarn.lock:6384]: Are we still using `lodash` in production indirectly?

Comment (viceice) [reply yarn.lock:6384]: image

yes, indirect by `yawn-yaml`

File: lib/workers/pr/index.ts (hunk 3 above)
Comment (viceice) [reply lib/workers/pr/index.ts:526]: good idea

File: yarn.lock (hunk 4 above)
Comment (rarkins) [reply yarn.lock:6384]: image

**Comment by JamieMagee** [comment]
@rarkins Did something change in logger? I didn't change anything there, but all the lint errors are around `traverse` in logger.

**Comment by rarkins** [comment]
@JamieMagee not that I can recall. Not according to GitHub either: https://github.com/renovatebot/renovate/commits/master/lib/logger

**Comment by JamieMagee** [comment]
It's because I created a dependency from `tools` to `lib`. Tools is compiled with the default `tsconfig.json` (really just `tsc -c tools`) so it type checks a lot of `lib` and fails as a result.

**Comment by JamieMagee** [comment]
Needs coverage fixes.

**Comment by viceice** [comment]
> It's because I created a dependency from `tools` to `lib`. Tools is compiled with the default `tsconfig.json` (really just `tsc -c tools`) so it type checks a lot of `lib` and fails as a result.

?? There should be no dep from tools to lib. Tools folder has its own tsconfig which should be used to compile only files in tools.

**Comment by JamieMagee** [comment]
> ?? There should be no dep from tools to lib.

True, but I accidentally added one when refactoring this code and moving the `camelCase` function to `lib/util/strings`. I've reverted this now.

File: lib/datasource/index.ts [hunk 5]
```diff
@@ -201,7 +201,7 @@ async function fetchReleases(
     }
     logError(datasource.id, config.lookupName, err);
   }
-  if (!dep || _.isEqual(dep, { releases: [] })) {
+  if (!dep || equal(dep, { releases: [] })) {
```
Comment (viceice) [inline lib/datasource/index.ts:204]: @rarkins Do we really need a deep equal here? Can't we simply check for `!dep?.releases?.length` ?

**Review by viceice — COMMENTED** [review]
Just one question, otherwise LGTM

File: lib/datasource/index.ts (hunk 5 above)
Comment (rarkins) [reply lib/datasource/index.ts:204]: No, there's a case for Go where it has empty releases but a valid sourceUrl. We later generate digests. So it's intentionally checking that only a truly empty object gets removed.

**Review by viceice — APPROVED** [review]

**Pull request merged**